code/modules/experisci/experiment/handlers/experiment_handler.dm

```diff
@@ -0,0 +1,393 @@
+	var/config_flags
+	/// Callback that, when supplied, can be called from the UI
+	var/datum/callback/start_experiment_callback
+
+/**
+ * Initializes a new instance of the experiment_handler component
+ *
+ * Arguments:
+ * * allowed_experiments - The list of /datum/experiment types that can be performed with this component
+ * * blacklisted_experiments - The list of /datum/experiment types that explicitly cannot be performed with this component
+ * * config_mode - The define that determines how the experiment_handler should display the configuration UI
+ * * disallowed_traits - Flags that control what experiment traits are blacklisted by this experiment handler
+ * * config_flags - Flags that control the operational behaviour of the experiment handler, see experiment defines
+ * * start_experiment_callback - When provided adds a UI button to use this callback to the start the experiment
+ */
+/datum/component/experiment_handler/Initialize(allowed_experiments = list(),
+												blacklisted_experiments = list(),
+												config_mode = EXPERIMENT_CONFIG_ATTACKSELF,
+												disallowed_traits = null,
+												config_flags = null,
+												var/datum/callback/start_experiment_callback = null)
+	. = ..()
+	if(!ismovable(parent))
+		return COMPONENT_INCOMPATIBLE
+	src.allowed_experiments = allowed_experiments
+	src.blacklisted_experiments = blacklisted_experiments
+	src.disallowed_traits = disallowed_traits
+	src.config_flags = config_flags
+	src.start_experiment_callback = start_experiment_callback
+
+	if(isitem(parent))
+		RegisterSignal(parent, COMSIG_ITEM_PRE_ATTACK, .proc/try_run_handheld_experiment)
+		RegisterSignal(parent, COMSIG_ITEM_AFTERATTACK, .proc/ignored_handheld_experiment_attempt)
+	if(istype(parent, /obj/machinery/doppler_array))
+		RegisterSignal(parent, COMSIG_DOPPLER_ARRAY_EXPLOSION_DETECTED, .proc/try_run_doppler_experiment)
+	if(istype(parent, /obj/machinery/destructive_scanner))
+		RegisterSignal(parent, COMSIG_MACHINERY_DESTRUCTIVE_SCAN, .proc/try_run_destructive_experiment)
+
+	// Determine UI display mode
+	switch(config_mode)
+		if(EXPERIMENT_CONFIG_ATTACKSELF)
+			RegisterSignal(parent, COMSIG_ITEM_ATTACK_SELF, .proc/configure_experiment)
+		if(EXPERIMENT_CONFIG_ALTCLICK)
+			RegisterSignal(parent, COMSIG_CLICK_ALT, .proc/configure_experiment)
+		if(EXPERIMENT_CONFIG_CLICK)
+			RegisterSignal(parent, COMSIG_ATOM_UI_INTERACT, .proc/configure_experiment_click)
+		if(EXPERIMENT_CONFIG_UI)
+			RegisterSignal(parent, COMSIG_UI_ACT, .proc/ui_handle_experiment)
+
+	// Auto connect to the first visible techweb (useful for always active handlers)
+	// Note this won't work at the moment for non-machines that have been included
+	// on the map as the servers aren't initialized when the non-machines are initializing
+	if (!(config_flags & EXPERIMENT_CONFIG_NO_AUTOCONNECT))
+		var/list/found_servers = get_available_servers(parent)
+		var/obj/machinery/rnd/server/s = found_servers.len ? found_servers[1] : null
```

Comment:
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names